frb_codegen/src/ir/ty_enum.rs

```diff
@@ -31,14 +31,22 @@ impl IrTypeTrait for IrTypeEnumRef {
     fn dart_api_type(&self) -> String {
         self.name.to_string()
     }
-    fn dart_wire_type(&self) -> String {
-        self.rust_wire_type()
+    fn dart_wire_type(&self, wasm: bool) -> String {
+        if wasm {
+            "List<dynamic>".into()
+        } else {
+            self.rust_wire_type(wasm)
+        }
     }
     fn rust_api_type(&self) -> String {
         self.name.to_string()
     }
-    fn rust_wire_type(&self) -> String {
-        format!("wire_{}", self.name)
+    fn rust_wire_type(&self, wasm: bool) -> String {
+        if wasm {
+            "Box<[JsValue]>".into()
+        } else {
+            format!("wire_{}", self.name)
+        }
```

Comment:
Seems that we are doing quite different things (very different types) here for wasm and plain-old cases.

Instead of "if", shall we make two different `IrTypeEnumRef` structs/implementations? When there is something in common, it can be extracted as common code.

Reply:
There's a small problem with that though -- by making WASM-ness part of IrType one needs to parse the program twice: once with and once without WASM. In the interest of code hygiene, we could find a different entry point for context-sensitive code, for example an *adapter* pattern of sorts.

Reply:
>  by making WASM-ness part of IrType one needs to parse the program twice

Oops you are right. I did not realize it is an *ir* (originally thought it is generator).

> In the interest of code hygiene, we could find a different entry point for context-sensitive code, for example an adapter pattern of sorts.

Totally agree! Happy to see that you also know design patterns and work hard on code quality. I have not digested the whole PR so surely my understanding of how to refactor to improve code quality is not as deep as you :)